### ISSUE TITLE
Installation via `setup.py` is depreciated

### ISSUE THREAD

**Issue by stggh** [PyAbel/PyAbel#371]
Hi @DanHickstein and @MikhailRyazanov (a belated happy new year!) - I have been playing with my new [Orangepi5]( http://www.orangepi.org/html/hardWare/computerAndMicrocontrollers/details/Orange-Pi-5.html) and was testing Anaconda python on it. However, the installation of `PyAbel` complains about the use of `setuptools`, which appears to be depreciated:
```python
(py310) [orangepi5:.+e/Python/PyAbel]$ python setup.py install --user
================================================================================
Warning: Cython extensions will not be built as Cython is not installed!
         This means that the abel.direct C implementation will not be available.
================================================================================
running install
/usr/lib/python3/dist-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
  warnings.warn(
/usr/lib/python3/dist-packages/setuptools/command/easy_install.py:158: EasyInstallDeprecationWarning: easy_install command is deprecated. Use build and pip and other standards-based tools.
```
I wonder what is an alternative installation method, apart from `pip`?   I haven't yet searched for a solution - using you guys as a short-cut!

FYI PyAbel benchmarks (default OS had older versions of `setuptools`), Dell has mkl libraries installed:
```
Dell Optiplex                                          Orange Pi 5                                             Raspberry pi 4B       
i7-9700 CPU @ 3.00GHz                                  Cortex-A76/A55 @ 2.4GHz                                 Cortex-A72 @ 1.5GHz       
PyAbel benchmark run on x86_64                         PyAbel benchmark run on aarch64                         PyAbel benchmark run on       
                                                                                                                
time in milliseconds                                   time in milliseconds                                    time in milliseconds       
                                                                                                                
=== Basis generation ==============================    === Basis generation ==============================     === Basis generation ==============================
                                                                                                                
Method              n = 501    n = 1001    n = 2001    Method              n = 501    n = 1001    n = 2001     Method              n = 501    n = 1001    n = 2001
---------------------------------------------------    ---------------------------------------------------     ---------------------------------------------------
basex                 572.4      3466.8     23908.4    basex                1014.0      6031.0     40596.5     basex                4706.2     30988.6    225344.7
daun                    1.9         4.2         9.5    daun                    3.5         7.9        19.3     daun                   17.6        46.6       122.4
linbasex               12.0        40.0       171.1    linbasex               25.2       111.5       449.0     linbasex              169.7       613.4      2471.5
onion_peeling           1.6         4.0        18.9    onion_peeling           4.3        21.5       145.4     onion_peeling          38.9       346.9      2719.5
rbasex                  7.0        16.1        50.5    rbasex                 14.0        41.1       173.8     rbasex                 78.1       306.6      1708.2
three_point             3.9        13.5        63.2    three_point            10.1        40.0       171.6     three_point            52.2       427.0      2074.7
two_point               1.2         4.3        20.1    two_point               2.5        10.6        42.9     two_point              12.7       108.2       479.9
                                                                                                                
                                                                                                                
=== Forward Abel transform ========================    === Forward Abel transform ========================     === Forward Abel transform ========================
                                                                                                                
Method              n = 501    n = 1001    n = 2001    Method              n = 501    n = 1001    n = 2001     Method              n = 501    n = 1001    n = 2001
---------------------------------------------------    ---------------------------------------------------     ---------------------------------------------------
basex                   0.2         4.9        57.9    basex                   2.3        16.7       127.6     basex                  70.8       638.0      5020.7
basex(var.reg.)         3.8        19.2       105.6    basex(var.reg.)         9.9        58.9       413.8     basex(var.reg.)       178.0      1581.5     12257.4
daun                    0.3         1.8        14.0    daun                    2.4        17.1       129.4     daun                   74.4       651.1      5086.9
direct_C               10.2        68.3       518.0                                                             
direct_Python         160.2      1135.2     12098.1    direct_Python         352.3      2986.9     24171.2     direct_Python        4021.5     57374.2    446129.0
hansenlaw               7.5        28.7       100.2    hansenlaw              11.1        46.1       227.6     hansenlaw              65.1       262.3      2008.7
rbasex                  1.8         7.6        32.0    rbasex                  3.4        14.2        54.1     rbasex                 26.3       198.4       770.2
rbasex(None)            0.9         4.3        17.8    rbasex(None)            1.8         7.2        27.4     rbasex(None)            8.4        59.0       227.2
                                                                                                                
                                                                                                                
 === Inverse Abel transform ========================   === Inverse Abel transform ========================     === Inverse Abel transform ========================
                                                                                                                
 Method              n = 501    n = 1001    n = 2001   Method              n = 501    n = 1001    n = 2001     Method              n = 501    n = 1001    n = 2001
 ---------------------------------------------------   ---------------------------------------------------     ---------------------------------------------------
 basex                   0.2         1.5        13.2   basex                   2.4        16.7       127.3     basex                  70.1       644.0      5022.3
 basex(var.reg.)         2.7        11.4        51.4   basex(var.reg.)        10.0        59.1       414.3     basex(var.reg.)       192.5      1694.9     13349.0
 daun                    0.3         1.8        13.0   daun                    2.2        13.2        86.1     daun                   55.3       536.3      3763.8
 daun(var.reg.)          1.6         6.4        39.0   daun(var.reg.)          8.0        48.8       354.2     daun(var.reg.)        177.8      1791.9     20631.5
 direct_C               10.9        74.9       520.9                                                           
 direct_Python         170.1      1138.1     11991.7   direct_Python         363.1      3022.3     24247.4     direct_Python        4106.8     57972.4    452701.8
 hansenlaw               8.6        30.7       109.6   hansenlaw              11.9        48.1       246.9     hansenlaw              69.2       278.4      2047.0
 linbasex               83.9       328.9      2231.6   linbasex              259.1      1435.2      8680.8     linbasex             1637.6     10379.6     74991.0
 onion_bordas          214.0      1041.4     11151.7   onion_bordas          426.1      2262.4     16941.7     onion_bordas         2144.1     11407.2     79856.4
 onion_peeling           0.2         1.6        13.0   onion_peeling           2.4        16.8       129.4     onion_peeling          70.6       636.5      5022.9
 rbasex                  1.6         7.2        32.4   rbasex                  3.4        14.2        54.0     rbasex                 26.3       196.7       768.9
 rbasex(None)            1.0         4.2        17.4   rbasex(None)            1.8         7.2        27.3     rbasex(None)            8.4        59.9       227.2
 three_point             0.2         1.4        12.8   three_point             2.4        17.0       140.2     three_point            70.3       638.7      5122.6
 two_point               0.2         1.6        13.0   two_point               2.5        16.8       141.5     two_point              67.0       639.0      5140.4

```
Steve

**Comment by stggh** [comment]
This web page has an indepth analysis [Summary](https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html#summary) (I cheated and went straight to the Summary). `pip install -e <dir path>` appears to be enough.
```python
(py310) [orangepi5:.+e/Python/PyAbel]$ pip install -e ./
Obtaining file:///home/stg121/Software/Python/PyAbel
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Requirement already satisfied: setuptools>=44.0 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (59.6.0)
Requirement already satisfied: six>=1.10.0 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.16.0)
Requirement already satisfied: scipy>=1.2 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.8.0)
Requirement already satisfied: numpy>=1.16 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.21.5)
Building wheels for collected packages: PyAbel
  Building editable for PyAbel (pyproject.toml) ... done
  Created wheel for PyAbel: filename=PyAbel-0.9.0-0.editable-py3-none-any.whl size=8014 sha256=811336fc12dfce7d145ddade42f46f02ae381ff78863501bd5f19fe52687e379
  Stored in directory: /tmp/pip-ephem-wheel-cache-23_bf4yt/wheels/b9/8f/bd/17a7a600167c4f9f1b6d7118701459ca6ed5cd5d477bf0f2e8
Successfully built PyAbel
Installing collected packages: PyAbel
  Attempting uninstall: PyAbel
    Found existing installation: PyAbel 0.9.0
    Uninstalling PyAbel-0.9.0:
      Successfully uninstalled PyAbel-0.9.0
Successfully installed PyAbel-0.9.0
```

**Comment by DanHickstein** [comment]
Happy New Year to you too!

Thanks for pointing this out. It does seem like a good idea to switch our recommended installation method to whatever is currently considered best practice. What didn't work about the pip install?

**Comment by stggh** [comment]
Thanks for the late night reply Dan. 

It seem(s) my problem(s) related to Anaconda python 3.10.  Anaconda python 3.11 does install using `setup.py` and `pip install ./`.  (see below)
 
```python
(py310) [orangepi5:.+e/Python/PyAbel]$ pip install ./
Processing /home/stg121/Software/Python/PyAbel
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy>=1.16 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.21.5)
Requirement already satisfied: six>=1.10.0 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.16.0)
Requirement already satisfied: setuptools>=44.0 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (59.6.0)
Requirement already satisfied: scipy>=1.2 in /usr/lib/python3/dist-packages (from PyAbel==0.9.0) (1.8.0)
Building wheels for collected packages: PyAbel
  Building wheel for PyAbel (pyproject.toml) ... done
  Created wheel for PyAbel: filename=PyAbel-0.9.0-py3-none-any.whl size=155108 sha256=a5dc1652a2b24705de8109fb874651e7ff9f2b70d054a3471218e6ccc4d3d590
  Stored in directory: /tmp/pip-ephem-wheel-cache-a8svj053/wheels/b9/8f/bd/17a7a600167c4f9f1b6d7118701459ca6ed5cd5d477bf0f2e8
Successfully built PyAbel
Installing collected packages: PyAbel
  Attempting uninstall: PyAbel
    Found existing installation: PyAbel 0.9.0
    Uninstalling PyAbel-0.9.0:
      Successfully uninstalled PyAbel-0.9.0
Successfully installed PyAbel-0.9.0

(py310) [orangepi5:.+e/Python/PyAbel]$ pd ~/
(py310) [orangepi5:/home/stg121]$ python
Python 3.10.9 (main, Jan 11 2023, 15:15:32) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import abel
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/stg121/anaconda3/envs/py310/lib/python3.10/site-packages/abel/__init__.py", line 25, in <module>
    from . import basex
  File "/home/stg121/anaconda3/envs/py310/lib/python3.10/site-packages/abel/basex.py", line 18, in <module>
    from abel.tools.polynomial import PiecewisePolynomial
  File "/home/stg121/anaconda3/envs/py310/lib/python3.10/site-packages/abel/tools/__init__.py", line 12, in <module>
    from . import symmetry
  File "/home/stg121/anaconda3/envs/py310/lib/python3.10/site-packages/abel/tools/symmetry.py", line 10, in <module>
    from scipy import fftpack
  File "/usr/lib/python3/dist-packages/scipy/fftpack/__init__.py", line 94, in <module>
    from ._basic import *
  File "/usr/lib/python3/dist-packages/scipy/fftpack/_basic.py", line 8, in <module>
    from scipy.fft import _pocketfft
  File "/usr/lib/python3/dist-packages/scipy/fft/__init__.py", line 91, in <module>
    from ._helper import next_fast_len
  File "/usr/lib/python3/dist-packages/scipy/fft/_helper.py", line 3, in <module>
    from ._pocketfft import helper as _helper
  File "/usr/lib/python3/dist-packages/scipy/fft/_pocketfft/__init__.py", line 3, in <module>
    from .basic import *
  File "/usr/lib/python3/dist-packages/scipy/fft/_pocketfft/basic.py", line 6, in <module>
    from . import pypocketfft as pfft
ImportError: /home/stg121/anaconda3/envs/py310/bin/../lib/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /usr/lib/python3/dist-packages/scipy/fft/_pocketfft/pypocketfft.cpython-310-aarch64-linux-gnu.so)
>>> quit()
```

3.11: Is fine:
```python
(py311) [orangepi5:.+e/Python/PyAbel]$ pip install ./
Processing /home/stg121/Software/Python/PyAbel
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy>=1.16 in /home/stg121/.local/lib/python3.11/site-packages/numpy-1.24.1-py3.11-linux-aarch64.egg (from PyAbel==0.9.0) (1.24.1)
Requirement already satisfied: setuptools>=44.0 in /home/stg121/anaconda3/envs/py311/lib/python3.11/site-packages (from PyAbel==0.9.0) (65.5.0)
Requirement already satisfied: scipy>=1.2 in /home/stg121/.local/lib/python3.11/site-packages/scipy-1.10.0-py3.11-linux-aarch64.egg (from PyAbel==0.9.0) (1.10.0)
Requirement already satisfied: six>=1.10.0 in /home/stg121/.local/lib/python3.11/site-packages/six-1.16.0-py3.11.egg (from PyAbel==0.9.0) (1.16.0)
Building wheels for collected packages: PyAbel
  Building wheel for PyAbel (pyproject.toml) ... done
  Created wheel for PyAbel: filename=PyAbel-0.9.0-py3-none-any.whl size=155108 sha256=93614f004340a703aff42422ff12ae2a446fabb6aa6b8c5c87ed90e52c3b075a
  Stored in directory: /tmp/pip-ephem-wheel-cache-n77t44s_/wheels/93/11/90/86ef91e48d6594dd6abcf4f1da76730eb95503cb5c7c8e35c1
Successfully built PyAbel
Installing collected packages: PyAbel
  Attempting uninstall: PyAbel
    Found existing installation: PyAbel 0.9.0
    Uninstalling PyAbel-0.9.0:
      Successfully uninstalled PyAbel-0.9.0
Successfully installed PyAbel-0.9.0
(py311) [orangepi5:.+e/Python/PyAbel]$ pd
py311) [orangepi5:/home/stg121]$ python
Python 3.11.0 (main, Jan 16 2023, 14:11:53) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import abel
>>> abel.benchmark.AbelTiming()
n = 301
  basex
    estimated ???, actually 1.054 s
  direct_Python
    estimated ???, actually 0.181 s
  onion_peeling
    estimated ???, actually 0.005 s
      :
PyAbel benchmark run on aarch64

time in milliseconds

=== Basis generation ==================

Method              n = 301     n = 501
---------------------------------------
basex                 633.0      1270.4
daun                    5.6         4.1
linbasex                9.7        25.9
onion_peeling           3.6        20.8
rbasex                  9.0       102.3
three_point             4.8        10.4
two_point               2.5         2.6


=== Forward Abel transform ============

Method              n = 301     n = 501
---------------------------------------
basex                   0.6         2.4
basex(var.reg.)       210.1       205.4
daun                    0.6         2.5
direct_Python          89.5       371.9
hansenlaw               5.6        12.2
rbasex                  4.7         3.5
rbasex(None)            2.5         1.8


=== Inverse Abel transform ============

Method              n = 301     n = 501
---------------------------------------
basex                   0.6         2.3
basex(var.reg.)       210.2       245.3
daun                    0.7         2.4
daun(var.reg.)        232.6       225.8
direct_Python          91.2       373.3
hansenlaw               6.1        14.4
linbasex               81.9       254.9
onion_bordas          149.8       460.3
onion_peeling           1.3         2.4
rbasex                  6.1         3.5
rbasex(None)            2.5         1.8
three_point             0.6         2.4
two_point               0.6         2.4
```

The remaining issue, is whether `setup.py` is depreciated (and as an aside, whether Python 3.11 is indeed faster than <= 3.10)

**Comment by MikhailRyazanov** [comment]
Happy new year to all!

I would agree that declarative package description should be cleaner than an imperative `setup.py`, but I have no idea how to do this with Cython extensions (moreover, doing so optionally). [Cython documentation](https://cython.readthedocs.io/en/stable/src/quickstart/build.html) itself only says: “Write a distutils/setuptools `setup.py`. This is the normal and recommended way.” and lists several other ways, which are no better...

Since we plan to move from Travis CI to GitHub actions for testing and deployment (see #365), maybe there are already some “standard” solutions for this? What other projects using Cython do?

**Comment by MikhailRyazanov** [comment]
I've looked at some other projects and read some documentation, and it seems that:
1. For example, NumPy, SciPy and Matplotlib still use non-trivial `setup.py`, so it's probably unavoidable.
2. According to current [setuptools docs](https://setuptools.pypa.io/en/latest/userguide/quickstart.html#setuppy-discouraged), running `python setup.py` _directly_ is deprecated, but using the `setup.py` file itself for configuration is supported (although they recommend to “expose as much as possible configuration in a more declarative way via the `pyproject.toml` or `setup.cfg`, and keep the `setup.py` minimal with only the dynamic parts”).
3. The recommended installation is [apparently](https://setuptools.pypa.io/en/latest/deprecated/commands.html) using
   ```shell
   pip install .  # replacement for python setup.py install
   pip install --editable .  # replacement for python setup.py develop
   ```
Thus, we probably need to update “Installation” in our README according to this. Can somebody check that `pip install .` works in Python 2 as well? If yes, we should just remove the subtitles “With pip” and “With setuptools”, and replace `python setup.py <...>` with these new `pip install [-e] .` commands.

Splitting the configuration into `setup.py`/`setup.cftg`/`pyproject.toml` is probably not very meaningful, I think.

**Comment by stggh** [comment]
Thanks for the excellent assessment @MikhailRyazanov!

I can confirm that `pip install .` works under python2:
```python
(base) [etalon:.+e/Python/PyAbel]$ git clone https://github.com/PyAbel/PyAbel.git
Cloning into 'PyAbel'...
remote: Enumerating objects: 8677, done.
remote: Counting objects: 100% (1493/1493), done.
remote: Compressing objects: 100% (503/503), done.
remote: Total 8677 (delta 1008), reused 1345 (delta 957), pack-reused 7184
Receiving objects: 100% (8677/8677), 22.78 MiB | 17.58 MiB/s, done.
Resolving deltas: 100% (6218/6218), done.
P(base) [etalon:.+e/Python/PyAbel]$ cd PyAbel
(base) [etalon:.+n/PyAbel/PyAbel]$ pip2 install .
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support pip 21.0 will remove support for this functionality.
Defaulting to user installation because normal site-packages is not writeable
Processing /home/stg121/Software/Python/PyAbel/PyAbel
Requirement already satisfied: numpy>=1.16 in /home/stg121/.local/lib/python2.7/site-packages (from PyAbel==0.9.0) (1.16.6)
Requirement already satisfied: scipy>=1.2 in /home/stg121/.local/lib/python2.7/site-packages (from PyAbel==0.9.0) (1.2.3)
Requirement already satisfied: setuptools>=44.0 in /usr/lib/python2.7/dist-packages (from PyAbel==0.9.0) (44.1.1)
Requirement already satisfied: six>=1.10.0 in /usr/local/lib/python2.7/dist-packages (from PyAbel==0.9.0) (1.10.0)
Building wheels for collected packages: PyAbel
  Building wheel for PyAbel (setup.py) ... done
  Created wheel for PyAbel: filename=PyAbel-0.9.0-cp27-cp27mu-linux_x86_64.whl size=515604 sha256=5584c15ca723c748890e4080a3302492cdf37872ab068d923c7e46e81e861779
  Stored in directory: /tmp/pip-ephem-wheel-cache-d_Ep75/wheels/cd/81/f4/e217ef4b70e0fedbcc200830632d3332eed9d820a70168b018
Successfully built PyAbel
Installing collected packages: PyAbel
  Attempting uninstall: PyAbel
    Found existing installation: PyAbel 0.8.0
    Uninstalling PyAbel-0.8.0:
      Successfully uninstalled PyAbel-0.8.0
Successfully installed PyAbel-0.9.0
(base) [etalon:.+n/PyAbel/PyAbel]$ cd ..

(base) [etalon:.+e/Python/PyAbel]$ python2 
Python 2.7.18 (default, Jul  1 2022, 11:13:17) 
[GCC 11.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import abel
>>> abel.benchmark.AbelTiming()
n = 301
  three_point
    estimated ???, actually 0.003 s
      :

PyAbel benchmark run on x86_64

time in milliseconds

=== Basis generation ==================

Method              n = 301     n = 501
---------------------------------------
basex                 269.2       810.0
daun                    2.7         3.3
linbasex                7.5        18.6
onion_peeling           2.5         5.0
rbasex                  8.7        18.4
three_point             3.0         6.8
two_point               0.6         1.5


=== Forward Abel transform ============

Method              n = 301     n = 501
---------------------------------------
basex                   0.3         0.8
basex(var.reg.)        85.1       115.4
daun                    0.4         1.0
direct_C                4.9        14.5
direct_Python          51.4       209.7
hansenlaw               3.2         7.8
rbasex                  1.3         3.5
rbasex(None)            0.7         2.0


=== Inverse Abel transform ============

Method              n = 301     n = 501
---------------------------------------
basex                   0.2         0.7
basex(var.reg.)         9.4        11.5
daun                    0.3         0.9
daun(var.reg.)         45.4        16.9
direct_C                7.2        27.1
direct_Python          88.6       247.9
hansenlaw               5.1        11.4
linbasex               71.9       129.6
onion_bordas          125.6       298.6
onion_peeling           0.2         0.8
rbasex                  1.2         3.3
rbasex(None)            0.7         2.0
three_point             0.2         0.5
two_point               0.2         0.5
>>>

**Comment by MikhailRyazanov** [comment]
Thanks, @stggh!
I thought that we would also need to update Travis and AppVeyor files to use `pip install .`, but they already use it! So this method not only works but in fact is tested. :–)

**Comment by MikhailRyazanov** [comment]
Off-topic: Orange Pi 5 looks serious. Does it really work without a heatsink?

**Comment by stggh** [comment]
Good value. I am still testing. Heat sink arriving in a few days:
 
![gnome-shell-screenshot-92rea8](https://user-images.githubusercontent.com/10932229/214175445-df354285-c861-4d39-9c38-bc573585699b.png)

**Comment by stggh** [comment]
@MikhailRyazanov - I may have misinterpreted your question on the Orangepi5. It operates fine without a heatsink, I did not notice any throttling, in some code tests. However, more demanding computation may benefit from having the heatsink. The heatsink arrived yesterday, and appears to do its job, according to qualitative `sensors` temperature readings.

My am investigating using the orangepi5 as a (replacement) mythtv system (back and frontend).

**Comment by MikhailRyazanov** [comment]
And regarding Python 3.11/3.10 performance, I don't think that we'll notice any significant difference except maybe for `direct_Python`, since everything else spends most of the time in NumPy code...

**Comment by stggh** [comment]
Solved by PR #372
